X16 Reference - Appendix F - 65C816 Processor.md

```diff
@@ -0,0 +1,2470 @@
+| P         | Processor Status | The flags. |
+| PC        | Program Counter  | The address of the current CPU instruction |
+
+.A, .X, and .Y can be 8 or 16 bits wide, based on the flag settings (see below).
+
+The Stack Pointer (.S) is 16 bits wide in Native mode and 8 bits wide (and fixed
+to the $100-1FF range) in Emulation mode.
+
+.DB and .K are the bank registers, allowing programs and data to occupy separate
+64K banks on computers with more than 64K of RAM. (The X16 does not use the bank
+registers, instead using addresses $00 and $01 for banking.)
+
+## Status Flags
+
+The native mode flags are as follows:
+
+`nvmx dizc e`
+
+  n = Negative  
+  v = oVerflow  
+  m = Memory width (0=16 bit, 1=8 bit)  
+  x = Index register width (0=16 bit, 1=8 bit)  
+  d = Decimal Mode  
+  i = Interupts Disabled  
+  z = Zero  
+  c = Carry  
+  e = Emulation Mode (0=65C02 mode, 1=65C816 mode)
+
+In emulation mode, the **m** and **x** flags are always set to 1.
+
+Here are the 6502 and 65C02 registers, for comparison:
+
+`nv1b dizc e`
+
+  n = Negative  
+  v = oVerflow  
+  1 = this bit is always 1  
+  b = brk: set during a BRK instruction interrupt  
+  d = Decimal Mode  
+  i = Interupts Disabled  
+  z = Zero  
+  c = Carry  
+  e = Emulation Mode (0=65C02 mode, 1=65C816 mode)
+
+**e** can only accessed via the XCE instruction, which swaps Carry and
+the Emulation flag.
+
+The other flags can all be manipulated with SEP and REP, and the various
+branch instructions (BEQ, BCS, etc) test some of the flags. The rest
+can only be tested indirectly through the stack.
+
+When a BRK or IRQ is triggered in _emulation_ mode, a ghost **b** flag
+will be pushed to the stack instead of the **x** flag. This can be used
+to test for a BRK vs IRQ in the Interrupt handler.
+
+## 16 bit mode
+
+The 65C816 CPU boots up in emulation mode. This locks the register width to
+8 bits and locks out certain operations.
+
+If you want to use the '816 features, including 16-bit operation, you will
+need to enable _native_ mode. Clearing **e** switches the CPU to native mode.
+However, it's not as simple as just setting a flag. The **e** flag can only
+be accessed through the XCE instruction, which swaps the Carry and Emulation
+flags.
+
+To switch to native mode, use the following steps:
+
+```
+CLC  ; clear the Carry bit
+XCE  ; swap the Emulation and Carry bit
+```
+
+To switch back to emulation mode, _set_ the Carry flag and perform an XCE again.
+
+```
+SEC  ; Set Carry
+XCE  ; and push the 1 into the Emulation flag.
+```
+
+Once **e** is cleared, the **m** and **x** flags can be set to 1 or 0 to control
+the register width.
+
+When the **m** flag is *clear*, Accumulator operations and memory reads and writes
+will be 16-bit operations. The CPU reads two bytes at a time with LDA, writes
+two bytes at a time with STA, and all math involving .A is 16 bits wide.
```

Comment:
We have to refer to it as C when it's 16 bits wide since A and B still exist - they describe the low and high byte of C (see `xba`).

Reply:
It seems to only be used in 4 opcodes, and it seems like it was introduced mostly as a way to say "Hey, this is always a 16-bit value." 

Since every other place the accumulator is referenced, its size is based on **m**, I don't think .C is especially useful as a name, other than for those 4 instructions. I think using it regularly will just confuse people, especially beginners.

Reply:
I think it's a bit technical and slightly confusing. And naturally because instructions like `pha` operate on .C when m=0 rather than introducing a new mnemonic like `phc`.

But I think we should follow convention rather than trying to make things less confusing ourselves.  The accumulator is either .A or .C depending on mode.  We can tweak our wording later.